UnitTests/GitUITests/CancellationTokenSequenceTests.cs

```diff
@@ -0,0 +1,163 @@
+            var token1 = sequence.Next();
+
+            Assert.False(token1.IsCancellationRequested);
+
+            var token2 = sequence.Next();
+
+            Assert.True(token1.IsCancellationRequested);
+            Assert.False(token2.IsCancellationRequested);
+        }
+
+        [Test]
+        public void Next_throws_if_disposed()
+        {
+            var sequence = new CancellationTokenSequence();
+
+            sequence.Dispose();
+
+            Assert.Throws<ObjectDisposedException>(() => sequence.Next());
+        }
+
+        [Test]
+        public void Cancel_cancels_previous_token()
+        {
+            var sequence = new CancellationTokenSequence();
+
+            var token = sequence.Next();
+
+            Assert.False(token.IsCancellationRequested);
+
+            sequence.CancelCurrent();
+
+            Assert.True(token.IsCancellationRequested);
+        }
+
+        [Test]
+        public void Cancel_is_idempotent()
+        {
+            var sequence = new CancellationTokenSequence();
+
+            var token = sequence.Next();
+
+            Assert.False(token.IsCancellationRequested);
+
+            sequence.CancelCurrent();
+            sequence.CancelCurrent();
+            sequence.CancelCurrent();
+            sequence.CancelCurrent();
+        }
+
+        [Test]
+        public void Cancel_does_not_throw_if_no_token_yet_issued()
+        {
+            var sequence = new CancellationTokenSequence();
+
+            sequence.CancelCurrent();
+        }
+
+        [Test]
+        public void Dispose_cancels_previous_token()
+        {
+            var sequence = new CancellationTokenSequence();
+
+            var token = sequence.Next();
+
+            Assert.False(token.IsCancellationRequested);
+
+            sequence.Dispose();
+
+            Assert.True(token.IsCancellationRequested);
+        }
+
+        [Test]
+        public void Dispose_is_idempotent()
+        {
+            var sequence = new CancellationTokenSequence();
+
+            sequence.Next();
+
+            sequence.Dispose();
+            sequence.Dispose();
+            sequence.Dispose();
+            sequence.Dispose();
+        }
+
+        [Test]
+        public void Dispose_does_not_throw_if_no_token_yet_issued()
+        {
+            var sequence = new CancellationTokenSequence();
+
+            sequence.Dispose();
+        }
+
+        [Test]
+        public void Next_handles_concurrent_requests_correctly()
+        {
+            const int loopCount = 10000;
+
+            // Assume hyperthreading, so halve logical processors (could use WMI or P/Invoke for more robust answer)
+            var coreCount = Math.Max(1, Environment.ProcessorCount / 2);
+
+            if (coreCount == 1)
+            {
+                Assert.Inconclusive("This test requires more than one processor to run.");
+            }
+
+            using (var sequence = new CancellationTokenSequence())
+            using (var barrier = new Barrier(coreCount))
```

Comment:
I am not convinced that we benefit from introducing a barrier to the test.

I think that running them wild would result in more collisions:
```csharp
        [Test]
        public void Concurrent_requests_to_Next_cancel_all_tokens_but_one()
        {
            const int loopCount = 10000;

            var coreCount = Environment.ProcessorCount;

            // Assume hyperthreading, so halve logical processors (could use WMI or P/Invoke for more robust answer)
            if (coreCount <= 2)
            {
                Assert.Inconclusive("This test requires more than one processor to run.");
            }

            using (var sequence = new CancellationTokenSequence())
            using (var countdown = new CountdownEvent(coreCount))
            {
                var allTokens = new ConcurrentBag<CancellationToken>();
                var lastTokens = new ConcurrentBag<CancellationToken>();

                for (var i = 0; i < coreCount; i++)
                {
                    new Thread(ThreadMethod).Start();
                }

                Assert.True(countdown.Wait(TimeSpan.FromSeconds(10)), "Test should have completed within a reasonable amount of time");

                Assert.AreEqual(loopCount * coreCount, allTokens.Count);
                allTokens.Where(token => !token.IsCancellationRequested).Single();
                lastTokens.Where(token => !token.IsCancellationRequested).Single();

                return;

                void ThreadMethod()
                {
                    CancellationToken token;
                    for (int i = 0; i < loopCount; i++)
                    {
                        token = sequence.Next();
                        allTokens.Add(token);
                    }

                    lastTokens.Add(token);
                    countdown.Signal();
                }
            }
        }
```

Also this test runs 10x faster.

Reply:
These seem like two different tests. The goal of the current test is to validate each iteration in turn. The first barrier sets the up, ready to take their locks. The second barrier ensures everyone has a token, then they test that only one of them succeeded.

Actually technically it just assumes that the aggregate count is representative. It's possible that some iterations are doubled, and others skipped. In practice that's highly unlikely, but it is a hole in the test.

I'm curious about the distribution of 'winner' across threads. Will report back.

If I understand the test code above, it validates that all tokens are eventually cancelled, but doesn't make the same guarantees about when that happens. Perhaps there's room for both tests. Running them wild possible results in more variety in the concurrency.

Would be interested to hear @sharwell's perspective.

Reply:
> I'm curious about the distribution of 'winner' across threads.

Added a count and got:

> Win count by index: 2575,2363,2545,2517

It is practically uniform, which is great. I take this to mean that the race is very even and they have lots of opportunity to interact in the ways we want to test for.

Reply:
>but doesn't make the same guarantees about when that happens.

There is already a test that assures that Next_cancels_previous_token. If `Next` has failed to cancel the prev token, then there is no way the prev token will get canceled later. I think that it is enough  to ensure that we don't leak tokens not being canceled. I also ensure that the not canceled token is among the tokens generated as the last token in each thread.

Reply:
I'm fine with neither/either/both, provided the outcome meets these:

1. The tests are reliable (i.e. won't fail on CI sporadically)
2. The tests aren't long enough to where they cause an observable increase in build times